### DISCUSSION TITLE
Pass through everything but note on/note off

### DISCUSSION THREAD

**Discussion by fapplin** [FortySevenEffects/arduino_midi_library#332] [Q&A]
I have a micro-controller that I'm receiving MIDI on via a 5-DIN connector from my keyboard controller. I would like to handle only note on/note off and pass through all other messages to the software synth. Is there an easy way to do this or do I have handle all messages via callbacks? And, if I have to do callbacks can you give me an example of using callbacks in a class?

Thanks.

**Comment by franky47** [comment]
That would be very easy to with the upcoming `filter` feature:

```c++
// Return `true` to let the message Thru, false to consume it:
bool everythingButNotes(const midi::Message& message)
{
  return message.type != midi::NoteOn && message.type != midi::NoteOff;
}

// The usual callbacks to handle Note messages:
void onNoteOn(byte pitch, byte velocity, byte channel)
{
}

void onNoteOff(byte pitch, byte velocity, byte channel)
{
}

void setup()
{
  MIDI.setThruFilter(everythingButNotes);
  MIDI.setHandleNoteOn(onNoteOn);
  MIDI.setHandleNoteOff(onNoteOff);
  // ...
}
```

I'm hoping to get it released sometime in September, watch this space!

**Reply by fapplin** [reply]
That looks great!! So, just out of curiosity - without the filter would I have to handle every callback?

**Reply by franky47** [reply]
Yes, and forward all other messages, which is verbose but doable in the mean time.

**Reply by fapplin** [reply]
Thanks and thanks for a great library!

**Reply by fapplin** [reply]
Has the setThruFilter() been implemented, yet? If so, can you tell me how to implement the callbacks in a class? I have a class called MyMidi that currently I wrap the sendNoteOn() and sendNoteOff() functions - like this.

void MyMidi::doNoteOn(byte channel, byte pitch, byte velocity) {
   MIDI.sendNoteOn(pitch, velocity, channel);
}

How can I change that so I'm using the callbacks like MIDI.setHandleNoteOn(onNoteOn), but via my class?

Thanks,

Frank

**Reply by franky47** [reply]
It's implemented in the `v5.1.0` branch, so you could pull the code out of that and use it.

Regarding interfacing with class methods, it depends on how you use instances of said class. Is it a singleton, or do you have multiple instances to bind to different (or same) callbacks & send methods?

**Reply by fapplin** [reply]
I'm not sure if showing the code helps answer your question, but here's a couple snippets.

In the main .ino file,  it looks like:

```
myMidi kmidi;

void setup() {
   kmidi.initialize();
   :
   :
}
```

in my MidiProcessing.cpp file,

```
#include "MidiProcessing.h"

// USB MIDI object
Adafruit_USBD_MIDI usb_midi;
//
// Create a new instance of the Arduino MIDI Library,
// and attach usb_midi as the transport.
MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, MIDI);


myMidi::myMidi()
{

} // end constructor

void myMidi:::initialize()
{
#if defined(ARDUINO_ARCH_MBED) && defined(ARDUINO_ARCH_RP2040)
  // Manual begin() is required on core without built-in support for TinyUSB such as mbed rp2040
  TinyUSB_Device_Init(0);
#endif

  MIDI.begin(MIDI_CHANNEL_OMNI);
  MIDI.turnThruOff();

  Serial1.setRX(17);
  Serial1.setTX(16);
  Serial1.begin(31250);

  while ( !TinyUSBDevice.mounted() ) delay(1);

}
```

**Reply by fapplin** [reply]
OK. Looking at my code - I have 2 different things going on with the MIDI. I'm reading in the MIDI via serial, making modifications, and sending it out via USB. So, I think I will separate the 2 out. The MIDI  out through the USB - I will use straight C with the callbacks (because I'll only have 3 callbacks to do). The MIDI in through the serial port -  I'll handle in the MIDIProcessing class.

So,  the my question becomes how do I do the filtering.  Back in August you gave this proposed example.

```
// Return `true` to let the message Thru, false to consume it:
bool everythingButNotes(const midi::Message& message)
{
  return message.type != midi::NoteOn && message.type != midi::NoteOff;
}

// The usual callbacks to handle Note messages:
void onNoteOn(byte pitch, byte velocity, byte channel)
{
}

void onNoteOff(byte pitch, byte velocity, byte channel)
{
}

void setup()
{
  MIDI.setThruFilter(everythingButNotes);
  MIDI.setHandleNoteOn(onNoteOn);
  MIDI.setHandleNoteOff(onNoteOff);
  // ...
}
```

Is this how to do it now with your new version?

**Reply by franky47** [reply]
Yes, this should do it.
